crates/typst/src/foundations/calc.rs

```diff
@@ -459,6 +465,149 @@ pub fn ln(
+}
+
+#[func(title = "Bitwise OR")]
+/// Calculates the bitwise OR between two integers.
+///
+/// For the purposes of this function, the parameters are treated as signed
+/// integers of 64 bits.
+///
+/// ```example
+/// #calc.bit-or(64, 32)
+/// ```
+pub fn bit_or(
+    /// The left-hand operand of the bitwise OR.
+    lhs: i64,
+    /// The right-hand operand of the bitwise OR.
+    rhs: i64,
+) -> i64 {
+    lhs | rhs
+}
+
+#[func(title = "Bitwise XOR")]
+/// Calculates the bitwise XOR between two integers.
+///
+/// For the purposes of this function, the parameters are treated as signed
+/// integers of 64 bits.
+///
+/// ```example
+/// #calc.bit-xor(64, 96)
+/// ```
+pub fn bit_xor(
+    /// The left-hand operand of the bitwise XOR.
+    lhs: i64,
+    /// The right-hand operand of the bitwise XOR.
+    rhs: i64,
+) -> i64 {
+    lhs ^ rhs
+}
+
+#[func(title = "Bitwise Left Shift")]
+/// Shifts the operand's bits to the left by the specified amount.
+///
+/// For the purposes of this function, the operand is treated as a signed
+/// integer of 64 bits. An error will occur if the result is too large to
+/// fit in a 64-bit integer.
+///
+/// ```example
+/// #calc.bit-lshift(33, 2)
+/// #calc.bit-lshift(-1, 3)
+/// ```
+pub fn bit_lshift(
+    /// The operand whose bits will be shifted.
+    operand: i64,
+
+    /// The amount of bits to shift. Must not be negative.
+    shift: u32,
+) -> StrResult<i64> {
+    Ok(operand.checked_shl(shift).ok_or("the result is too large")?)
+}
+
+#[func(title = "Bitwise Right Shift")]
+/// Shifts the operand's bits to the right by the specified amount.
+/// Performs an arithmetic shift by default (extends the sign bit to the left,
+/// such that negative numbers stay negative), but that can be changed by the
+/// `logical` parameter.
+///
+/// For the purposes of this function, the operand is treated as a signed
+/// integer of 64 bits.
+///
+/// ```example
+/// #calc.bit-rshift(64, 2)
+/// #calc.bit-rshift(-8, 2)
+/// #calc.bit-rshift(-8, 2, logical: true)
+/// ```
+pub fn bit_rshift(
+    /// The operand whose bits will be shifted.
+    operand: i64,
+
+    /// The amount of bits to shift. Must not be negative.
+    /// This is capped at the operand's bits minus 1 (63).
+    /// Therefore, the shift will always succeed.
+    shift: u32,
+
+    /// Toggles whether a logical (unsigned) right shift should be performed
+    /// instead of arithmetic right shift (the default).
+    /// If this is `true`, negative operands will not preserve their sign bit,
+    /// and bits which appear to the left after the shift will be 0.
+    /// This parameter has no effect on non-negative operands.
+    #[named]
+    #[default(false)]
+    logical: bool,
+) -> i64 {
+    // Don't panic on excessive shift.
+    let shift = shift.min(i64::BITS - 1);
+
+    if !logical || operand >= 0 {
+        // For non-negative operands, arithmetic and logical shifts produce the
+        // same results.
+        operand >> shift
```

Comment:
I think the `operand >= 0` condition is unnecessary and confusing when reading the code. I'd suggest simply performing the unsigned shift when `logical` is true, no matter what.

Reply:
When I was writing this for the first time, I wanted to avoid unnecessary extra logic. But turns out that the extra logic is just casting, which is a no-op, so it should be fine to remove this check.

Reply:
Done, thanks for the suggestion!